sklearn/feature_extraction/tests/test_text.py

```diff
@@ -1196,3 +1197,51 @@ def build_preprocessor(self):
                                             .findall(doc),
                     stop_words=['and'])
     assert _check_stop_words_consistency(vec) is True
+
+
+@pytest.mark.parametrize('Estimator',
+                         [CountVectorizer, TfidfVectorizer, HashingVectorizer])
+def test_callable_analyzer_vs_file_input(Estimator):
+    data = ['this is text, not file or filename']
+
+    with pytest.raises(FileNotFoundError):
+        Estimator(analyzer=lambda x: x.split(),
+                  input='filename').fit_transform(data)
+
+    with pytest.raises(AttributeError,
+                       match="'str' object has no attribute 'read'"):
+        Estimator(analyzer=lambda x: x.split(),
+                  input='file').fit_transform(data)
+
+    # check if the ChangedBehaviorWarning is raised if the given analyzer
+    # expects a file or a file name.
+    def analyzer1(doc):
+        open(doc, 'r')
+
+    def analyzer2(doc):
+        doc.read()
+
+    for analyzer in [analyzer1, analyzer2]:
+        for input_type in ['file', 'filename']:
+            try:
+                with pytest.warns(ChangedBehaviorWarning,
+                                  match="Since v0.21, vectorizer"):
+                    Estimator(analyzer=analyzer,
+                              input=input_type).fit_transform(data)
+            except Exception:
+                pass
```

Comment:
To check that no warnings are raised maybe better to do https://stackoverflow.com/a/45671804

Reply:
I'm not sure what you mean, which warning would be raised here? I'm making sure that the right warning is raised and then ignore the exception which is raised after the warning being raised.

Reply:
> I'm making sure that the right warning is raised and then ignore the exception

OK, I though you were doing something else.

Still, if the expected warning is not emitted, `pytest.warns` will raise an `AssertionError`, which you would then catch and silently ignore. So basically this will never fail, won't it?

Reply:
Fixed by Guillaume's suggestion.